Sources/Basics/Environment/Environment.swift

```diff
@@ -0,0 +1,326 @@
+                throw UpdateEnvironmentError(
+                    function: "_putenv",
+                    code: Int32(GetLastError()))
+            }
+            #else
+            guard setenv(key.rawValue, value, 1) == 0 else {
+                throw UpdateEnvironmentError(
+                    function: "setenv",
+                    code: errno)
+            }
+            #endif
+        } else {
+            #if os(Windows)
+            guard _putenv("\(key)=") == 0 else {
+                throw UpdateEnvironmentError(
+                    function: "_putenv",
+                    code: Int32(GetLastError()))
+            }
+            #else
+            guard unsetenv(key.rawValue) == 0 else {
+                throw UpdateEnvironmentError(
+                    function: "unsetenv",
+                    code: errno)
+            }
+            #endif
+        }
+        Self._cachedCurrent.withLock { $0 = nil }
+    }
+}
+
+// MARK: - Cachable Keys
+extension Environment {
+    /// Returns a copy of `self` with known non-cacheable keys removed.
+    ///
+    /// - Issue: rdar://107029374
+    package var cachable: Environment {
+        var cachable = Environment()
+        for (key, value) in self {
+            if !EnvironmentKey.nonCachable.contains(key) {
+                cachable[key] = value
+            }
+        }
+        return cachable
+    }
+}
+
+// MARK: - Protocol Conformances
+extension Environment: Collection {
+    public struct Index: Comparable {
+        public static func < (lhs: Self, rhs: Self) -> Bool {
+            lhs.underlying < rhs.underlying
+        }
+
+        var underlying: Dictionary<EnvironmentKey, String>.Index
+    }
+    public typealias Element = (key: EnvironmentKey, value: String)
+    // FIXME: Remove after upgrading past Swift 5.9
+    // Required to be explicitly spelled out on older Swift compilers.
+    public typealias Iterator = IndexingIterator<Self>
+
+    public var startIndex: Index {
+        Index(underlying: self.storage.startIndex)
+    }
+
+    public var endIndex: Index {
+        Index(underlying: self.storage.endIndex)
+    }
+
+    public subscript(index: Index) -> Iterator.Element {
+        get { self.storage[index.underlying] }
+    }
+
+    public func index(after index: Self.Index) -> Self.Index {
+        Index(underlying: self.storage.index(after: index.underlying))
+    }
+}
+
+extension Environment: CustomStringConvertible {
+    public var description: String {
+        var description = "["
+        let sorted = self.sorted { $0.key < $1.key }
+        var first = true
+        for (key, value) in sorted {
+            if first {
+                first = false
+            } else {
+                description.append(", ")
+            }
+            description.append("\"")
+            description.append(key.rawValue)
+            description.append("=")
+            description.append(value)
+            description.append("\"")
+        }
```

Comment:
```suggestion
        sorted.map { #""\#($0.rawValue)"="\#($1)""# }.joined(", ")
```

Reply:
I avoided string interpolation and join to avoid intermediate allocations

Reply:
That said it really doesn't matter and I can use an easier to read method

Reply:
Curious if we have issues on the compiler repo tracking possible optimizer work that hypothetically could get rid of intermediate allocations here?